README.md
npd6
====

Neighbor Proxy Daemon IPv6


https://github.com/maxchendt/npd6/releases/download/1.1.2/npd6_1.1.2_amd64.deb

To build deb package

    wget https://github.com/maxchendt/npd6/archive/1.1.2.tar.gz
    tar -xvpzf 1.1.2.tar.gz 
    cd npd6-1.1.2/
    sudo apt-get install --no-install-recommends  debhelper build-essential
    dpkg-buildpackage -b -us -uc -i
    cd ..
    sudo dpkg -i npd6_1.1.2_amd64.deb 




debian/changelog
npd6 (1.1.2) stable; urgency=medium

  * add debian/, add support for systemd

 -- CHEN Deng-Ta <no@no.org>  Thu, 25 Aug 2016 21:48:08 +0800


npd6 (1.1.0) trusty; urgency=medium

  * Consolidation into 1.1.0

 -- Sean Groarke <sgroarke@gmail.com>  Wed, 16 Jul 2014 16:13:30 +0200

npd6 (1.0.1) trusty; urgency=medium

  * Fix for bug 60

 -- Sean Groarke <sgroarke@gmail.com>  Wed, 16 Jul 2014 16:12:30 +0200

npd6 (1.0.1) quantal; urgency=low

  * Massage into Launchpad-suitable form

 -- Sean Groarke <sgroarke@gmail.com>  Fri, 04 Jan 2013 15:47:20 +0100

npd6 (1.0.0) quantal; urgency=low

  * Merge DEV to TRUNK - Lots of changes and fixes.
  * Version bump to 1.0.0

 -- Sean Groarke <sgroarke@gmail.com>  Thu, 03 Jan 2013 10:03:18 +0100

npd6 (0.9.99) quantal; urgency=low

  * BETA release! Use with caution!
  * Bug 48. Allow use of ANY prefix.
  * Bug 49. No limits interface numbers.
  * Bug 50. Masks on the prefixes.
  * Bug 51
  * Bug 51 and 53. Substantial changes to building and packaging.
  * Many minor fixes and debug changes.
  * Sample config file updated.
  * MAN pages updated.

 -- Sean Groarke <sgroarke@gmail.com>  Wed, 02 Jan 2013 16:17:11 +0100

npd6 (0.7.0) precise; urgency=low

    * Multi-interface support.
    * Minor bug fixes and doc changes..

 -- Sean Groarke <npd6project@gmail.com>  Mon, 14 Nov 2012 07:00:00 +0200

npd6 (0.6.0) precise; urgency=low

    * Bug fix for issue 42.
    * Feature fix for issue 43.

 -- Sean Groarke <npd6project@gmail.com>  Mon, 15 Oct 2012 07:00:00 +0200

npd6 (0.5.2) precise; urgency=low

    * Bug fixes for issue 38 & 40.

 -- Sean Groarke <npd6project@gmail.com>  Fri, 12 Oct 2012 07:00:00 +0200

npd6 (0.5.1) unstable; urgency=low

  * Packaging tidy-up.

 -- Sean Groarke <sgroarke@gmail.com>  Wed, 03 Oct 2012 10:37:23 +0200

npd6 (0.5.0) precise; urgency=low

  * Bump version 

 -- Craig Miskell <craig@catalyst.net.nz>  Mon, 27 Aug 2012 08:44:59 +1200

npd6 (0.4.6) precise; urgency=low

  * Initial package.

 -- Sean Groarke <npd6project@gmail.com>  Mon, 21 May 2012 14:30:48 +0200


main.c
/*
*   This software is Copyright 2011 by Sean Groarke <sgroarke@gmail.com>
*   All rights reserved.
*
*   This file is part of npd6.
*
*   npd6 is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   npd6 is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with npd6.  If not, see <http://www.gnu.org/licenses/>.
*/

/* $Id$
* $HeadURL$
*/

#include "includes.h"
#include "npd6.h"

char usage_str[] =
{
    "\n"
    "  -c, --config=PATH      Sets the config file.  Default is /etc/npd6.conf.\n"
    "  -d, --debug            Sets the normal debug level.\n"
    "  -D, --debug2           Sets full debug level. (Lots!)\n"
    "  -f, --foreground       Run in foreground, otherwise daemonize.\n"
    "  -h, --help             Show this help screen.\n"
    "  -l  --logfile=PATH     Sets the log file, else default to syslog. If \"-l -\" then stdout/stderr used.\n"
    "  -v, --version          Print the version and quit.\n"
};

struct option prog_opt[] =
{
    {"config",  required_argument, 0, 'c'},
    {"debug",   optional_argument, 0, 'd'},
    {"debug2",  optional_argument, 0, 'D'},
    {"help",    optional_argument, 0, 'h'},
    {"logfile", required_argument, 0, 'l'},
    {"version", optional_argument, 0, 'v'},
    {"foreground", optional_argument, 0, 'f'},
    {0, 0, 0, 0}
};

#define OPTIONS_STR "c:dDhl:vf"

struct Interface *IfaceList = NULL;

int main(int argc, char *argv[])
{
    char logfile[FILENAME_MAX] = "";
    int c, err, loop;
    
    // Default some globals
    strncpy(configfile, NPD6_CONF, FILENAME_MAX);
    daemonize=1;
    // Default black/whitelisting to OFF
    listType = NOLIST;
    // Default config file values as required
    naLinkOptFlag = 0;
    nsIgnoreLocal = 1;
    naRouter = 1;
    maxHops = MAXMAXHOPS;
    
    // Logging
    listLog=0;
    ralog=0;
    
    /* Global interface and prefix counts */
    interfaceCount = 0;
    prefixCount = 0;
    
    /* Parse the args */
    while ((c = getopt_long(argc, argv, OPTIONS_STR, prog_opt, NULL)) > 0)
    {
        if (c==-1)
            break;
        
        switch (c) {
            case 'c':
                strcpy(configfile, optarg);
                break;
            case 'l':
                strcpy(logfile, optarg);
                break;
            case 'd':
                debug=1;
                break;
            case 'D':
                debug=2;
                break;
            case 'f':
                daemonize=0;
                break;
            case 'v':
                showVersion();
                return 0;
                break;
            case 'h':
                showUsage();
                return 0;
                break;
        }
    }
    
    /* Sort out where to log */
    if ( strlen(logfile) )
    {
        if (strlen(logfile) == 1 && (logfile[0]='-')  )
            logging = USE_STD;
        else
            logging = USE_FILE;
    }
    else
    {
        logging = USE_SYSLOG;
    }
    
    /* Open the log and config*/
    if ( (logging == USE_FILE) && (openLog(logfile) < 0)  )
    {
        printf("Exiting. Error in setting up logging correctly.\n");
        exit (1);
    }
    flog(LOG_INFO, "*********************** npd6 *****************************");
    
    if ( readConfig(configfile) )
    {
        flog(LOG_ERR, "Error in config file: %s", configfile);
        return 1;
    } 
    
    err = init_sockets();
    if (err) {
        flog(LOG_ERR, "init_sockets: failed to initialise %d sockets.", err);
        exit(1);
    }
    
    /* Set allmulti on the interfaces */
    for (loop=0; loop<interfaceCount; loop++)
    {
        interfaces[loop].multiStatus = if_allmulti(interfaces[loop].nameStr, TRUE);
    }
    
    /* Seems like about the right time to daemonize (or not) */
    if (daemonize)
    {
        if (daemon(0, 0) < 0 )
        {
            flog(LOG_ERR, "Failed to daemonize. Error: %s", strerror(errno) );
            exit(1);
        }
    }
    
    /* Set up signal handlers */
    signal(SIGUSR1, usersignal);
    signal(SIGUSR2, usersignal);
    signal(SIGHUP, usersignal);
    signal(SIGINT, usersignal);
    signal(SIGTERM, usersignal);    // Typically used by init.d scripts
    
    /* And off we go... */
    dispatcher();
    
    flog(LOG_ERR, "Fell back out of dispatcher... This is impossible.");
    return 0;
}

void dispatcher(void)
{
    struct pollfd   fds[3];
    unsigned int    msglen;
    unsigned char   msgdata[MAX_MSG_SIZE * 2];
    int             rc;
    int             fdIdx;
    int             consecutivePollErrors = 0;
    
    // Each interface has 2 sockets, so we need to allocate for that + 1
//     fds = (struct pollfd *)calloc( (interfaceCount*2)+1, sizeof(struct pollfd) );
//     flog(LOG_DEBUG2, "Dynamically allocated %d bytes to the master FD array", 
//          (interfaceCount+1) * sizeof(struct pollfd) );
    
    // In the fds set, the first N positions are for the v6 sockets, the second N
    // are for the icmpv6 sockets.
//     for(fdIdx=0; fdIdx < interfaceCount; fdIdx++)
//     {
        // Packet socket
        fds[0].fd = sockPkt;
        fds[0].events = POLLIN;
        fds[0].revents = 0;
        
        // ICMP socket
        // We only bother with this as we get inbound junk on this socket 
        // (including RAs which we might actually care about now cf. bug 60)
        fds[1].fd = sockIcmp;
        fds[1].events = POLLIN;
        fds[1].revents = 0;
//     }
    
    // Tail it
    fds[2].fd = -1;
    fds[2].events = 0;
    fds[2].revents = 0;
    
    for (;;)
    {
        rc = poll(fds, 2, DISPATCH_TIMEOUT);
        
        if (rc > 0)
        {
            // Most likely event is a valid data item received.
            for (fdIdx=0; fdIdx < 2; fdIdx++)
            {
                if (fds[fdIdx].revents & POLLIN)
                {
                    // Was it a packet socket?
                    if(fdIdx == 0) {
                        consecutivePollErrors = 0;	// reset it
                        msglen = get_rx(sockPkt, msgdata);
                        // msglen is checked for sanity already within get_rx()
                        flog(LOG_DEBUG2, "For packet socket, get_rx() gave msg with len = %d", msglen);
                        processNS(fdIdx, msgdata, msglen);
                        continue;
                    }
                    // Or was it an ICMP socket?
                    if(fdIdx == 1) {
                        struct in6_addr icmp6Addr;
                        consecutivePollErrors = 0;	// reset it
                        msglen = get_rx_icmp6(sockIcmp, msgdata, &icmp6Addr);
                        flog(LOG_DEBUG2, "For ICMP6 socket, get_rx_icmp6() gave msg with len = %d", msglen);
                        // We do nothing at all with the received data!
                        // Or maybe we do.... Ref. bug/NFR 60: process them
                        // and yank out the RAs for logging.
                        // Decide what to do based upon config file option ralog
                        if (ralog) 
                        {
                            processICMP(fdIdx, msgdata, msglen, &icmp6Addr);
                        }
                        continue;
                    }
                }
                
                // If it wasn't a POLLIN, it is likely an significant error
                if (fds[fdIdx].revents & (POLLERR | POLLHUP | POLLNVAL) )
                {
                    flog(LOG_WARNING, "Major socket error on fds %d", fdIdx);
                    // Try and recover... long shot
                    close(sockPkt);
                    sleep(1);
                    sockPkt = open_packet_socket();
                    if ( sockPkt < 0)
                    {
                        // Drop dead. We're stuffed.
                        flog(LOG_ERR, "dispatcher(): failed to reinit stuck socket. Dead.");
                        exit(1);
                    }
                    fds[fdIdx].fd = sockPkt;
                    fds[fdIdx].events = POLLIN;
                    fds[fdIdx].revents = 0;
                    
                    // Have we had more consecutive errors than the threshold value?
                    consecutivePollErrors++;
                    if ((pollErrorLimit > 0) && (consecutivePollErrors >= pollErrorLimit) )
                    {
                        flog(LOG_ERR, "dispatcher(): %d consecutive major poll errors. Terminating.",
                             consecutivePollErrors);
                        dropdead();
                    }
                    
                    continue;
                }
            }
            continue;
        }
        else if ( rc == 0 )
        {
            flog(LOG_DEBUG, "Stale select - Idling....... Low activity........");
            consecutivePollErrors = 0; // Using the select timeout as our quantum of error counting
            // Timer fired?
            // One day. If we implement timers.
        }
        else if ( rc == -1 )
        {
            /* Truly an error or maybe we processed a signal?*/
            if ( errno == EINTR )
            {
                flog(LOG_ERR, "Broke out of the poll() via a signal event.");
            }
            else
            {
                flog(LOG_ERR, "Weird poll error: %s", strerror(errno));
            }
            continue;
        }
        
        flog(LOG_DEBUG, "Timed out of poll(). Timeout was %d ms", DISPATCH_TIMEOUT);
    }
}


//*******************************************************
// Give advice.
void showUsage(void)
{
    fprintf(stderr, "usage: %s %s\n", pname, usage_str);
}


util.c
/*
 *   This software is Copyright 2011 by Sean Groarke <sgroarke@gmail.com>
 *   All rights reserved.
 *
 *   This file is part of npd6.
 *
 *   npd6 is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   npd6 is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with npd6.  If not, see <http://www.gnu.org/licenses/>.
 */

/* $Id$
 * $HeadURL$
 */

#include "includes.h"
#include "npd6.h"

//*******************************************************
// When we receive sigusrN, do something awesome.
/*****************************************************************************
 * usersignal
 *      When dispatcher picks up a user signal, do something with it.
 *
 * Inputs:
 *  int mysig
 *      The signal received.
 *
 * Outputs:
 *  Various, depending upon the sig.
 *
 * Return:
 *      void
 *
 */
void usersignal(int mysig)
{
    switch(mysig) {
        case SIGUSR1:
            signal(SIGUSR1, usersignal);
            flog(LOG_DEBUG, "called with USR1");
            flog(LOG_INFO, "SIGUSR1 received: rereading config");
            if ( readConfig(configfile) )
            {
                flog(LOG_ERR, "Error in config file: %s", configfile);
                exit(1);
            }
            break;
         case SIGUSR2:
            signal(SIGUSR2, usersignal);
            flog(LOG_DEBUG, "called with USR2");
            dumpAddressData();
            break;
        case SIGHUP:
            signal(SIGUSR2, usersignal);
            flog(LOG_DEBUG, "called with HUP");
            /* action? */
            break;
        case SIGINT:
        case SIGTERM:
            signal(SIGUSR2, usersignal);
            flog(LOG_DEBUG, "called with INT");
            /* We're dying, so handle directly here*/
            dropdead();
            break;
         default:
            flog(LOG_DEBUG, "Why am I here? Confused.");
            break;
    }
}



/*****************************************************************************
 * npd6log
 *      Log as per level to the previously defined log file or system..
 *
 * Inputs:
 *  char * fn name
 *      From the caller's __FUNCTION__
 *  int pri
 *      Log level: e.g. LOG_ERR, LOG_INFO, LOG_DEBUG, etc.
 *  char *format
 *      Standard format string.
 *  ...
 *      Variable number of params to accompany the format string.
 * GLOBAL
 *  logFileFD
 *      Previously opened lof device.
 *
 * Outputs:
 *  Via fprintf to the log file.
 *
 * Return:
 *      0 on success
 *
 * Notes:
 *      This will be called from the macro expansion of "flog()"
 *      This gets called a lot, but much of the time only does
 *      anything if the debug options are turned on. Hence it's
 *      written in a slightly odd manner to avoid data allocation
 *      or work unless we have debug turned on of it's a non-debug
 *      call.
 */
int npd6log(const char *function, int pri, char *format, ...)
{
    // Pick up debug requests and decide if we are logging them
    if (!debug && pri>=LOG_DEBUG)
    {
        return 0;  // Silent...
    }

    // Check for extra super power debug
    if ( (debug!=2) && (pri == LOG_DEBUG2) )
    {
        return 0;   // Silent...
    }

    // Normalise the debug level
    if( pri==LOG_DEBUG2)
    {
        pri=LOG_DEBUG;  // Since only we understand the two levels
    }

    // Artificial blocking of code to improve efficiency... if the compiler plays ball. :-)
    {
        time_t now;
        struct tm *timenow;
        char timestamp[128], obuff[2048];
        va_list param;

        va_start(param, format);
        vsnprintf(obuff, sizeof(obuff), format, param);
        now = time(NULL);
        timenow = localtime(&now);
        (void) strftime(timestamp, sizeof(timestamp), LOGTIMEFORMAT, timenow);

        switch (logging) {
            case USE_FILE:
                fprintf(logFileFD, "[%s] %s: %s\n", timestamp, function, obuff);
                break;
            case USE_SYSLOG:
                syslog(pri, "%s: %s\n", function, obuff);
                break;
            case USE_STD:
                if (pri <= LOG_ERR)
                    fprintf(stderr, "[%s] %s: %s\n", timestamp, function, obuff);
                else
                    fprintf(stdout, "[%s] %s: %s\n", timestamp, function, obuff);
                break;
        }

        va_end(param);

        return 0;
    }
}


/*****************************************************************************
 * print_addr
 *      Convert ipv6 address to string representation.
 *
 * Inputs:
 *  const struct in6_addr * addr
 *      Binary ipv6 address
 *
 * Outputs:
 *  char * str
 *      String representation, *not* fully padded.
 *
 * Return:
 *      void
 *
 * Notes:
 *  Compare with print_addr16 - this version does not pad.
 */
void print_addr(struct in6_addr *addr, char *str)
{
        const char *res;

        res = inet_ntop(AF_INET6, (void *)addr, str, INET6_ADDRSTRLEN);

        if (res == NULL)
        {
                flog(LOG_ERR, "print_addr: inet_ntop: %s", strerror(errno));
                strcpy(str, "[invalid address]");
        }
}


/*****************************************************************************
 * build_addr
 *      Convert char represetation of ipv6 addr to binary form.
 *
 * Inputs:
 *  char * str
 *      String representation, fully padded.
 *
 * Outputs:
 *  const struct in6_addr * addr
 *      Binary ipv6 address
 *
 * Return:
 *      return code from inet_pton
 *      1 => Good, else Bad
 */
int build_addr(char *str, struct in6_addr *addr)
{
    int ret;
    flog(LOG_DEBUG2, "called with address %s", str);
    ret = inet_pton(AF_INET6, str, (void *)addr);
    if (ret == 1)
        flog(LOG_DEBUG2, "inet_pton OK");
    else if(ret == 0)
        flog(LOG_ERR, "invalid input address");
    else
        flog(LOG_ERR, "inet_pton: %s", strerror(errno));

    return ret;
}



/*****************************************************************************
 * print_addr16
 *      Print ipv6 address in fully expanded 64-bit char form
 *
 * Inputs:
 *  const struct in6_addr * addr
 *      Binary ipv6 address
 *
 * Outputs:
 *  char * str
 *      String representation, fully padded.
 *
 * Return:
 *      void
 */
void print_addr16(const struct in6_addr * addr, char * str)
{
   sprintf(str, "%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x",
                 (int)addr->s6_addr[0], (int)addr->s6_addr[1],
                 (int)addr->s6_addr[2], (int)addr->s6_addr[3],
                 (int)addr->s6_addr[4], (int)addr->s6_addr[5],
                 (int)addr->s6_addr[6], (int)addr->s6_addr[7],
                 (int)addr->s6_addr[8], (int)addr->s6_addr[9],
                 (int)addr->s6_addr[10], (int)addr->s6_addr[11],
                 (int)addr->s6_addr[12], (int)addr->s6_addr[13],
                 (int)addr->s6_addr[14], (int)addr->s6_addr[15]);
}



/*****************************************************************************
 * prefixset
 *      Take prefix in the form 1111:2222:3333:4444: and pad it to the
 *      full length
 *
 * Inputs:
 *  char * px
 *      String representation, fully padded.
 *
 * Outputs:
 *  As input
 *
 * Return:
 *      -1 on error, else bit length of unpadded prefix.
 *
 * Note:
 *  IMPORTANT! px *must* be big enough to hold full ipv6 string
 */
int prefixset(char px[])
{
    size_t len;
    int missing, c1, c2;
    char suffix[INET6_ADDRSTRLEN];

    // First we must ensure fully padded with leading 0s
    for(c1=0; c1<INET6_ADDRSTRLEN; c1+=5 )
    {
        len = strlen(px);
        for(c2 = 0; c2 < 4; c2++)
        {
            if (px[c1+c2] != ':')
                continue;
            else
                break;
        }

        missing = abs(c2-4);
        if ( (missing>=1) && (missing<=3) )
        {
            strcpy( suffix, &px[c1]);       // Grab the tail
            memset(&px[c1], '0', missing);  // pad it with missing zeros
            px[c1+missing] = '\0';          // Re-terminate
            strcat(px, suffix);             // Add the tail back
        }
    }
    flog(LOG_DEBUG2, "String after padding: %s", px);
    
    // Do we have a '::' on the end we need to trim?
    len = strlen(px);
    if ( (px[len-1] == ':') && (px[len-2] == ':') )
    {
        flog(LOG_DEBUG2, "Spotted double :: termination of prefix.");
        px[len-1] = '\0';
    }
    
    len = strlen(px);
    switch (len) {
        case 5:
            strcat(px, "0000:0000:0000:0000:0000:0000:0000");
            return 16;
        case 10:
            strcat(px, "0000:0000:0000:0000:0000:0000");
            return 32;
        case 15:
            strcat(px, "0000:0000:0000:0000:0000");
            return 48;
        case 20:
            strcat(px, "0000:0000:0000:0000");
            return 64;
        case 25:
            strcat(px, "0000:0000:0000");
            return 80;
        case 30:
            strcat(px, "0000:0000");
            return 96;
        case 35:
            strcat(px, "0000");
            return 112;
        case 39:
            flog(LOG_ERR, "Full 128-bits defined as the configured prefix. Sure???");
            return 128;
        default:
            flog(LOG_ERR, "configured prefix not correctly formatted (len = %d)", len);
            return -1;
    }
}



/*****************************************************************************
 * stripwhitespace
 *  Tidy up lines of text from the config file.
 *
 * Inputs:
 *  char * str
 *      String to be tidied.
 *
 * Outputs:
 *  As input
 *
 * Return:
 *  void
 */
void stripwhitespace(char *str)
{
    char *p1 = str;
    char *p2 = str;

    p1=str;
    while(*p1 != 0) {
        if(isspace(*p1))
            ++p1;
        else
            *p2++ = *p1++;
    }
    *p2=0;
}



/*****************************************************************************
 * dumpHex
 *  Take a lump of binary data (like an ethernet frame!) and print it in hex.
 *  Useful for debugging!
 *
 * Inputs:
 *  unsigned char *data
 *      Lump of data
 *  unsigned int len
 *      Amount of data
 *
 * Outputs:
 *  Via printf to stdio
 *
 * Return:
 *  void
 */
void dumpHex(unsigned char *data, unsigned int len)
{
    int ix;

    printf("Dumping %d bytes of hex:\n>>>>>", len);

    for(ix=0; ix < len; ix++)
        printf("%02x", data[ix]);
    printf("<<<<<\n");
}


/*****************************************************************************
 * getLinkaddress
 *  Get the link-level address (i.e. MAC address) for an interface.
 *
 * Inputs:
 *  char * iface
 *      String containing the interface name (e.g. "eth1")
 *
 * Outputs:
 *  unsigned char * link
 *      String of the form "00:12:34:56:78:90"
 *
 * Return:
 *  0 is successful,
 *  1 if error.
 */
int getLinkaddress( char * iface, unsigned char * link) {
    int sockfd, io;
    struct ifreq ifr;

    strncpy( ifr.ifr_name, iface, INTERFACE_STRLEN );

    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if(sockfd < 0) {
        flog(LOG_ERR, "failed to open a test SOCK_STREAM.");
        return 1;
    }

    io = ioctl(sockfd, SIOCGIFHWADDR, (char *)&ifr);
    if(io < 0) {
        flog(LOG_ERR, "octl failed to obtain link address");
        return 1;
    }

    memcpy(link, (unsigned char *)ifr.ifr_ifru.ifru_hwaddr.sa_data, 6);

    close(sockfd);

    return 0;
}


//*******************************************************
// Take the supplied filename and open it for logging use.
// Upon return, logFileFD set unless we failed.
int openLog(char *logFileName)
{
    if (logging == USE_FILE)
    {
        if ((logFileFD = fopen(logFileName, "a")) == NULL)
        {
            fprintf(stderr, "Can't open %s: %s\n", logFileName, strerror(errno));
            return (-1);
        }
        // Set line-buffering so we don't need to explicitly fflush()
        // when we write via npd6log
        setlinebuf(logFileFD);
        return 0;
    }

    if (logging == USE_SYSLOG)
    {
        openlog("npd6", (LOG_NDELAY|LOG_PID),LOG_DAEMON);
        return 0;
    }

    // Error
    return -1;
}


//*******************************************************
// Just display the version and return.
void showVersion(void)
{
    printf("npd6 - version %s\n", BUILDREV);
    printf("\nCopyright (C) 2011-2013 Sean Groarke\n\n");
}


void dropdead(void)
{
    int loop;
    
    /* We're dying, so tidy up*/
    /* Restore interface flags and close sockets */
    for (loop=0; loop<interfaceCount; loop++)
    {
        if_allmulti(interfaces[loop].nameStr, interfaces[loop].multiStatus);
        close( sockPkt );
        close( sockIcmp );
    }

    flog(LOG_ERR, "Tidied up and now exiting. Goodbye.");
    exit(0);
}


/*****************************************************************************
 * dumpData
 *  Dump internal data. Initially this will mean the set of collected
 *  target addresses seen (if that option is enabled)
 *
 * Inputs:
 *  tRoot is the tree of collected targets.
 *
 * Outputs:
 *  Data is dumped to the defined log.
 *
 * Return:
 *  Void
 */
void dumpAddressData(void)
{
    if (!collectTargets)
    {
        flog(LOG_INFO, "Not dumping collected addresses - feature disabled via config.");
        return;
    }

    flog(LOG_INFO, "====================================");
    flog(LOG_INFO, "Dumping list of targets seen so far:");
    flog(LOG_INFO, "------------------------------------");

    twalk(tRoot, tDump);

    if (tEntries == collectTargets)
    {
        flog(LOG_INFO, "(reached the configured limit - there were maybe more.)");
    }

    flog(LOG_INFO, "Total unique targets seen: %d", tEntries);
    flog(LOG_INFO, "====================================");
}


/*****************************************************************************
 * storeTarget
 *  Look in tRoot tree to see if we have it already. If we don't store
 *  it in the tree. If we do have it, then ignore.
 *
 * Inputs:
 *  in6_addr *Target - this is the newly seen target to check
 *
 * Outputs:
 *  tRoot has a new item added if the address was new.
 *
 * Return:
 *  Void
 */
void storeTarget(struct in6_addr *newTarget)
{
    struct in6_addr *ptr;

    // Take a permanent copy of the target
    ptr = (struct in6_addr *)malloc(sizeof(struct in6_addr) );
    if (!ptr)
    {
        flog(LOG_ERR, "Malloc failed. Ignoring.");
	return;
    }
    memcpy(ptr, newTarget, sizeof(struct in6_addr) );

    // We can't just tsearch() it into the tree, as there's no
    // way to then know if the entry already existed or is now newly created.
    // Hence we can't know to bump the count. So we tfind() first
    // and only tsearch() if required. In a typical net the initial
    // tfind() is going to return a result almost all the time, so the
    // overhead of this double-call is actually low.
    if ( tfind( (void *)ptr, &tRoot, tCompare) == NULL )
    {
        if (tEntries >= collectTargets)
        {
            flog(LOG_INFO, "Reached max threshold of recorded targets (%d). Not recording.", collectTargets);
            return;
        }
        // New entry
        flog(LOG_DEBUG2, "New entry - recording.");
        if ( tsearch( (void *)ptr, &tRoot, tCompare) == NULL)
        {
            flog(LOG_ERR, "tsearch failed. Cannot record entry.");
            return;
        }
        else
        {
            tEntries++;
        }
    }
    else
    {
        flog(LOG_DEBUG2, "Entry already recorded. Ignoring.");
    }
}

/*****************************************************************************
 * tCompare
 *  This is the compare fn used by the tree handler.
 *
 * Inputs:
 *  The two generic pointers are struct in6_addr *.
 *
 * Outputs:
 *  None.
 *
 * Return:
 *  0 if item already present, 1 if it was new.
 *
 * Notes:
 * Need to compare two 128 bit numbers! Yucky.
 * On 64-bit, we could assume long int => 64 bit, but
 * given we may be on 32-bit, need to assume long int
 * is max 32 bit, as per ANSI.
 *
 * We also make use of the underlying structure of in6_addr,
 * which is int[16]
 *
 * Needs to be moderately efficient, since if we're recording
 * a lot of addresses we call this via tfind/tsearch quite a lot,
 * so we do minimal-comparison.
 */
int tCompare(const void *pa, const void *pb)
{
    int paI=0, pbI=0;
    int idx;

    for(idx=0; idx<=15; idx++)
    {
        paI = ((struct in6_addr *)pa)->s6_addr[idx];
        pbI = ((struct in6_addr *)pb)->s6_addr[idx];

        if (paI == pbI)
            continue;
        if (paI < pbI)
            return -1;
        else
            return 1;
    };

    // If we reach here, the items were identical
    return 0;
}


/*****************************************************************************
 * tDump
 *  This is the action used when walking tRoot from dumpData()
 *
 * Inputs:
 *  node, type of visit, depth - Check the man page for more...
 *
 * Outputs:
 *  Data dumped to log.
 *
 * Return:
 *  void
 */
void tDump(const void *nodep, const VISIT which, const int depth)
{
    struct in6_addr *data;
    char addressString[INET6_ADDRSTRLEN];

    switch (which) {
        case preorder:
        case endorder:
            break;
        case postorder:
        case leaf:
            data = *(struct in6_addr **) nodep;
            print_addr(data, addressString);
            flog(LOG_INFO, "Address: %s", addressString);
            break;
    }
}


/*****************************************************************************
 * storeListEntry
 *
 * Inputs:
 *  in6_addr *Target - this is the newly seen target to check
 *
 * Outputs:
 *  lRoot has a new item added if the address was new.
 *
 * Return:
 *  Void
 */
void storeListEntry(struct in6_addr *newEntry)
{
    struct in6_addr *ptr;

    // Take a permanenet copy of the target
    ptr = (struct in6_addr *)malloc(sizeof(struct in6_addr) );
    if (!ptr)
    {
        flog(LOG_ERR, "Malloc failed. Ignoring.");
	return;
    }
    memcpy(ptr, newEntry, sizeof(struct in6_addr) );

    if ( tfind( (void *)ptr, &lRoot, tCompare) == NULL )
    {
        // New entry
        flog(LOG_DEBUG2, "New list entry");
        if ( tsearch( (void *)ptr, &lRoot, tCompare) == NULL)
        {
            flog(LOG_ERR, "tsearch failed. Cannot record entry.");
            return;
        }
    }
    else
    {
        flog(LOG_ERR, "Dupe list entry. Ignoring.");
    }
}





